text/0032-fog-compatible-gift-codes.md

```diff
@@ -0,0 +1,255 @@
+but this is not private, and the fog operator can see which `TxOut` you looked up. This puts the
+gift codes outside the normal threat-model for Fog transfers.
+
+We propose to change the design to improve this
+* Instead of sending the `tx_out_public_key`, we should send the global index of the tx out. Then, even a fog user
+  can use this to recover the `TxOut` privately from fog-ledger. Also this results in a slightly smaller payload,
+  and serves the purpose of uniquely identifying the `TxOut`.
+  
+Second, we propose an additional change in the design which improves some aspects of this.
+* Instead of sending money to a temporary account, we try to just transfer ownership of a specific `TxOut`, by giving them
+the private keys to that TxOut
+  * To enable a recipient to see the unblinded amount of the `TxOut`, we just have to give them the `shared_secret`.
+  * To enable a recipient to spend a `TxOut`, we just have to give them the `onetime_private_key`.
+  
+Given this data, a Fog user who receives the payload can ask Fog ledger for a merkle proof of membership for the `TxOut` of given index.
+They can get the entire `TxOut` this way, match it against the shared secret and so unblind the amount commitment, and given the one-time
+private key, they can spend it as normal using the transaction builder.
+
+A major concern for mobile apps is the ability to track things across linked devices using the blockchain (and not a separate synchronization
+mechanism). It is a concern for instance that one app may be trying to give away a `TxOut` in a gift code, while another selects it to be spent in a transaction.
+
+To avoid that, we propose to standardize that `GIFT_CODE_SUBADDRESS_INDEX = 2`, and that before building a gift code, the app should make a self-payment to this
+subaddress witht he correct amount for the gift code. Then the gift code is built against the `TxOut` that belongs to subaddress 2.
+
+The advantage of this is that unlike with the temporary address, the fog user is able to track the gift code.
+* The `TxOut` still belongs to the sender's address, even if we gave the private keys away to someone (who may not have used it yet).
+* We don't lose track of the in-flight gift code if the app is restarted -- all linked devices can learn about the in-flight gift code because of the use of subaddress 2.
+  This was not the case with the temporary address.
+* Because the `TxOut` still belongs to us, we can see if the gift code is actually used by the recipient, and if for some reason it is never used, we can
+  reclaim the money. (This would still be possible with the previous design, but it's more complicated because we have to preserve the temporary address entropy off-chain somewhere.)
+* The app can also more simply have recoverable transaction history around the use of gift codes, because they can represent in-bound `TxOut`'s to subaddress 2 as gift-code funding
```

Comment:
```suggestion
* The app can also more simply have recoverable transaction history around the use of gift codes, because they can represent in-bound `TxOut`'s to subaddress `u64::MAX - 1` as gift-code funding
```